src/saml.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the SamlApp class to encapsulate the business logic."""
+import logging
+import urllib
+from functools import cached_property
+from typing import List, Set, Optional
+
+from lxml import etree  # nosec
+
+from charm_state import CharmConfigInvalidError, CharmState
+
+from pydantic import AnyHttpUrl, BaseModel, Field
+
+logger = logging.getLogger(__name__)
+
+
+class SamlEndpoint(BaseModel):  # pylint: disable=too-few-public-methods
```

Comment:
Should this be in the state?

Reply:
The state doesn't hold data on the SamlEndpoints. This looks more cohesive to me

Reply:
@gregory-schiano any thoughts?

Reply:
I think we shouldn't pay too much attention for now, since there's no Provider library, which will change the whole shape of this project.
The goal here is to have a working charm implementing the relation to validate it on the relation repository and then start working on the integrator seriously

Reply:
I'm resolving the comment, then